### PR TITLE
golang swig support for swss-common, unit test included

### PR DESCRIPTION
**Pull request by jipanyang**
Signed-off-by: Jipan Yang <jipan.yang@alibaba-inc.com>

jipan@e92e7f4f35e8:/sonic/src/sonic-swss-common/tests$ GOPATH=/tmp/sonic/go go test -v redis_ut_test.go
=== RUN   Test_Table
--- PASS: Test_Table (0.00s)
=== RUN   Test_ProducerStateTable
--- PASS: Test_ProducerStateTable (0.00s)
=== RUN   Test_SubscriberStateTable
--- PASS: Test_SubscriberStateTable (0.07s)
PASS
ok  	command-line-arguments	0.071s

### CODE REVIEW

File: goext/swsscommon.i [hunk 1]
```diff
@@ -0,0 +1,78 @@
+%module swsscommon
```
Comment (lguohan) [inline goext/swsscommon.i]: are we duplicating this file for both python and go? it means we are duplicating the maintenance cost. I prefer to have one.

File: goext/Makefile.in [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+GOPATHDIR = $(GOPATH)/src/$(REPODIR)
+
+all: build
+
+build: $(SRCDIR_CXXSRCS)
+	# To use swig version 3.0.12 .
+	if [ ! -x $(SWIGEXE) ]; then \
+		git clone https://github.com/swig/swig.git; \
+		cd swig && git checkout rel-3.0.12 && ./autogen.sh && ./configure && make; cd - ; \
+	fi
```
Comment (lguohan) [inline goext/Makefile.in]: swig is installed in the sonic-slave, if you need a new version, we need to add a project to compile swig 3.0.12 in the buildimage and first build that and then build this project.

File: goext/swsscommon.i (hunk 1 above)
Comment (JipanYanga) [reply goext/swsscommon.i]: Thanks for the review!  

The problem I encountered with go is that "Selectable" argout could not be returned in $result as that in python which has PyObject to take extra variable, but in go it is just int type.  Also Selectable as go interface is not usable outside of swsscommon package.  

I'm thinking of adding void Select::isSelected(Selectable *selectable) method in Select c++ class to get around the issue, so in go after select(), we could check whether certain Select child interface has been selected.   Any other suggestion?  

Current subscribeStateTable unit test is skipping the argout of Selectable. 
int Select::select(Selectable **c, int *fd, unsigned int timeout)

File: goext/Makefile.in (hunk 2 above)
Comment (JipanYanga) [reply goext/Makefile.in]: Yes, existing swig doesn't support -cgo option which is required for go. Will try to see how to add a project for swig.

**Comment by JipanYanga** [comment]
The other problem I have is that "GOPATH = /sonic/go" is being used,  it works in my slave docker build environment, but it is failing in sonic jenkin build environment.  It looks it doesn't allow folder creation for /sonic/go/src/github.com/Azure/sonic-swss-common.

We'll need a GOPATH set,  other modules need the same environment variable to build and import swsscommon go package.     Any suggestion on this issue? Thanks!

mkdir -p /sonic/go/src/github.com/Azure/sonic-swss-common 2>/dev/null || true
rm -f -r /sonic/go/src/github.com/Azure/sonic-swss-common/*
cp swsscommon_wrap.cxx /sonic/go/src/github.com/Azure/sonic-swss-common/
cp: cannot create regular file '/sonic/go/src/github.com/Azure/sonic-swss-common/': No such file or directory

**Comment by lguohan** [comment]
is this needed?

**Comment by jipanyang** [comment]
This is change is not needed by current version of sonic-telemetry.   I'm thinking of closing this PR for now.

**Comment by jipanyang** [comment]
No use case for swss-common swig go support yet,  close it for now.
btw,  the way c++ mapped to golang in swig is awkward,  it works but requires extra care to use in native golang application.

**Pull request closed without merging**